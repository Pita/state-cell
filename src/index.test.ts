import { createStateCell, StateCellReducer } from '.';

interface Rectangle {
  width: number;
  height: number;
}

interface EnsureRatioMsg {
  ratio: number;
}

interface ScaleMsg {
  scaleFactor: 2;
}

test('basic state test', () => {
  const initialState: Rectangle = {
    width: 3,
    height: 4,
  };
  const scale: StateCellReducer<Rectangle, ScaleMsg> = (state, msg) => ({
    width: state.width * msg.scaleFactor,
    height: state.height * msg.scaleFactor,
  });
  const ensureRatio: StateCellReducer<Rectangle, EnsureRatioMsg> = (
    state,
    msg,
  ) => ({ width: state.width, height: state.width * msg.ratio });

  const msgs = { scale, ensureRatio };
  const stateCell = createStateCell<Rectangle, typeof msgs>(initialState, msgs);

  // original state
  expect(stateCell.state.width).toEqual(3);
  expect(stateCell.state.height).toEqual(4);

  // scale
  stateCell.msg.scale({ scaleFactor: 2 });
  expect(stateCell.state.width).toEqual(6);
  expect(stateCell.state.height).toEqual(8);

  // ratio
  stateCell.msg.ensureRatio({ ratio: 2 });
  expect(stateCell.state.width).toEqual(6);
  expect(stateCell.state.height).toEqual(12);
});

test('rejects invalid message names', () => {
  const initialState: Rectangle = {
    width: 3,
    height: 4,
  };

  const Scale: StateCellReducer<Rectangle, ScaleMsg> = (state, msg) => ({
    width: state.width * msg.scaleFactor,
    height: state.height * msg.scaleFactor,
  });

  const msgs = { Scale };
  expect(() => {
    createStateCell<Rectangle, typeof msgs>(initialState, msgs);
  }).toThrowErrorMatchingSnapshot();
});
