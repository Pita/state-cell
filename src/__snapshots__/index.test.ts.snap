// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejects invalid message names 1`] = `"'Scale' is not a valid message name. Message names must be alphanumeric and start with a lowercase letter"`;
